### PR TITLE
[FLINK-9563]: Using a custom sink function for tests in CEPITCase instead of writing to disk

### PR DESCRIPTION
**Pull request by deepaks4077**
## What is the purpose of the change

This change modifies the CEPITCase integration test to use a custom sink function to collect and compare test results, instead of writing them to a file. It does not add/remove any constituent tests.

## Brief change log

- Removed Before and After junit annotations
- Added a custom sink function with a static arraylist to collect and compare test results

## Verifying this change

This change is already covered by existing tests, such as CEPITCase, which is the end to end test of the CEP API.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by deepaks4077** [comment]
@zentol, thanks, will do!

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 1]
```diff
@@ -360,27 +345,30 @@ public boolean filter(Event value) throws Exception {
 		});
 
 		DataStream<String> result = CEP.pattern(input, pattern).select(
-			new PatternSelectFunction<Event, String>() {
+				new PatternSelectFunction<Event, String>() {
 
-				@Override
-				public String select(Map<String, List<Event>> pattern) {
-					StringBuilder builder = new StringBuilder();
+					@Override
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: indendation changes that weren't reversed

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 2]
```diff
@@ -504,20 +498,30 @@ public String select(Map<String, List<Event>> pattern) {
 					StringBuilder builder = new StringBuilder();
 
 					builder.append(pattern.get("start").get(0).getPrice()).append(",")
-						.append(pattern.get("middle").get(0).getPrice()).append(",")
-						.append(pattern.get("end").get(0).getPrice());
+							.append(pattern.get("middle").get(0).getPrice()).append(",")
+							.append(pattern.get("end").get(0).getPrice());
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: indendation change, revert like the others

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 3]
```diff
@@ -573,19 +577,25 @@ public String select(Map<String, List<Event>> pattern) {
 				StringBuilder builder = new StringBuilder();
 
 				builder.append(pattern.get("start").get(0).getId()).append(",")
-					.append(pattern.get("middle").get(0).getId()).append(",")
-					.append(pattern.get("end").get(0).getId());
+						.append(pattern.get("middle").get(0).getId()).append(",")
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: indendation

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 4]
```diff
@@ -708,10 +724,32 @@ public boolean filter(Tuple2<Integer, String> rec) throws Exception {
...
 		});
 
-		result.writeAsText(resultPath, FileSystem.WriteMode.OVERWRITE);
+		CollectSink.VALUES.clear();
 
-		expected = "(1,a)\n(3,a)";
+		result.map(new MapFunction<Tuple2<Integer, String>, String>() {
+			@Override
+			public String map(Tuple2<Integer, String> value) throws Exception {
+				return value.toString();
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: please do not compare as strings. This was only done for simplicity in the previous code so we don't have to re-parse the contents of the text file.

This also applies to other tests.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 5]
```diff
@@ -658,20 +668,26 @@ public String select(Map<String, List<Event>> pattern) {
 					StringBuilder builder = new StringBuilder();
 
 					builder.append(pattern.get("start").get(0).getId()).append(",")
-						.append(pattern.get("middle").get(0).getId()).append(",")
-						.append(pattern.get("end").get(0).getId());
+							.append(pattern.get("middle").get(0).getId()).append(",")
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: indentation

**Comment by zentol** [comment]
env.execute() should block until the job has finished execution, so your analysis may be incorrect.

I can take a deeper look at possible semantic changes, but there are some conceptual issues I'd like to clear up first.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 6]
```diff
@@ -708,10 +724,32 @@ public boolean filter(Tuple2<Integer, String> rec) throws Exception {
...
+
+		assertEquals(expected, CollectSink.VALUES);
 	}
+
+	private static class CollectSink implements SinkFunction<String> {
+
+		public static final List<String> VALUES = new ArrayList<>();
+
+		@Override
+		public synchronized void invoke(String value) throws Exception {
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: this synchronization may not work correctly if multiple instances of this function exists. Synchronize on `VALUES` instead.

**Comment by deepaks4077** [comment]
@zentol , thanks for your comments. I'll take a look soon!

**Comment by deepaks4077** [comment]
@zentol, I've updated the tests and all tests pass now. Please take a look when you get the chance.

**Comment by deepaks4077** [comment]
@zentol, do you have some time this week to review this? I can proceed with the other tests once I know that this is the right way to handle it :)

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 7]
```diff
@@ -229,7 +206,7 @@ public void testSimplePatternEventTime() throws Exception {
 			Tuple2.of(new Event(4, "end", 4.0), 10L),
 			Tuple2.of(new Event(5, "middle", 5.0), 7L),
 			// last element for high final watermark
-			Tuple2.of(new Event(5, "middle", 5.0), 100L)
+			Tuple2.of(new Event(6, "middle", 5.0), 100L)
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: why is this change necessary?

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java (hunk 5 above)
Comment (zentol) [reply flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: still needs to be reverted

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 8]
```diff
@@ -512,12 +491,16 @@ public String select(Map<String, List<Event>> pattern) {
...
+		List<Either<String, String>> resultList = new ArrayList<>();
+
+		DataStreamUtils.collect(result).forEachRemaining(resultList::add);
 
-		// the expected sequences of matching event ids
-		expected = "Left(1.0)\nLeft(2.0)\nLeft(2.0)\nRight(2.0,2.0,2.0)";
+		resultList.sort(Comparator.comparing(either -> either.toString()));
 
-		env.execute();
+		List<Either<String, String>> expected = Arrays.asList(Either.Left.of("1.0"), Either.Left.of("2.0"),
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: put each  element on a new line:
```
Arrays.asList(
	a,
	b
	c
```

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 9]
```diff
@@ -512,12 +491,16 @@ public String select(Map<String, List<Event>> pattern) {
...
+												Either.Left.of("2.0"), Either.Right.of("2.0,2.0,2.0"));
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: indentation should only be 1 tab

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 10]
```diff
@@ -580,12 +563,17 @@ public String select(Map<String, List<Event>> pattern) {
 			}
 		});
 
-		result.writeAsText(resultPath, FileSystem.WriteMode.OVERWRITE);
+		List<String> resultList = new ArrayList<>();
+
+		DataStreamUtils.collect(result).forEachRemaining(resultList::add);
+
+		List<String> expected = Arrays.asList("1,5,6\n1,2,3\n4,5,6\n1,2,6".split("\n"));
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: changes this line to work like other tests, inline `Arrays.asList` into the `assertEquals` call, split the list right away instead of with `split()`.

File: flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java [hunk 11]
```diff
@@ -666,12 +654,17 @@ public String select(Map<String, List<Event>> pattern) {
 			}
 		);
 
-		result.writeAsText(resultPath, FileSystem.WriteMode.OVERWRITE);
+		List<String> resultList = new ArrayList<>();
+
+		DataStreamUtils.collect(result).forEachRemaining(resultList::add);
+
+		List<String> expected = Arrays.asList("1,6,4\n1,5,4".split("\n"));
```
Comment (zentol) [inline flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/CEPITCase.java]: same as above

**Comment by deepaks4077** [comment]
@zentol, thanks for reviewing. I've updated the PR, can you please take a look?

**Comment by zentol** [comment]
alrighty, looks good to me, merging.
